# Лабораторная работа 1
### Python, Jupyther, Numpy, Pandas

## Numpy
![Image of numpy](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1a/NumPy_logo.svg/1200px-NumPy_logo.svg.png)

NumPy — библиотека с открытым исходным кодом для Python.
Возможности:
* поддержка многомерных массивов (включая матрицы);
* поддержка высокоуровневых математических функций, предназначенных для работы с многомерными массивами.

Документация: http://www.numpy.org/

In [2]:
import numpy as np

### Возможности

In [3]:
# Создание массива (ndarray) из списка (list)
list_1 = [0,1,2,3,4]
arr_1d = np.array(list_1)

# Сделайте print полученного массива и его типа
# Ваш код

In [4]:
# Прибавление и вычитание числа

for i in range(len(list_1)):
    list_1[i] += 1
    
arr_1d += 1

print(list_1)
print(arr_1d)

for i in range(len(list_1)):
    list_1[i] -= 1
    
arr_1d -= 1

print(list_1)
print(arr_1d)

# Также работает умножение и деление

[1, 2, 3, 4, 5]
[1 2 3 4 5]
[0, 1, 2, 3, 4]
[0 1 2 3 4]


In [5]:
# Массивы могут быть созданы различными способами
zeros = np.zeros(10)
ones = np.ones(10)
arange = np.arange(10)
linspace = np.linspace(0, 1, 5)

print('zeros', zeros)
print('ones', ones)
print('arange', arange)
print('linspace', linspace)

zeros [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
ones [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
arange [0 1 2 3 4 5 6 7 8 9]
linspace [0.   0.25 0.5  0.75 1.  ]


In [6]:
# Массивы могут складываться, вычитаться, умножаться, делится

print(ones+arange)
print(ones*arange)
print(ones/arange) # имеется возможность для деления на 0 - приводит к inf, отсутствующие значения к np.nan

[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
[       inf 1.         0.5        0.33333333 0.25       0.2
 0.16666667 0.14285714 0.125      0.11111111]


C:\Users\Acer\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: RuntimeWarning: divide by zero encountered in true_divide
  """


In [7]:
# Индексация и двумерные массивы
print(np.arange(10)[5:9])

arange = np.arange(10).reshape(2, 5) # команда reshape в данном случае позволяет получуть двумерный массив
print(arange)
print(arange[:, 3:])

# также можно проводить экстракцию по условию
print(arange[arange%2 == 1])

[5 6 7 8]
[[0 1 2 3 4]
 [5 6 7 8 9]]
[[3 4]
 [8 9]]
[1 3 5 7 9]


Гораздо больше возможностей на странице с документацией: http://www.numpy.org/

### Задачи

#### 1. Создайте одномерный массив, в котором по порядку все числа до 100, кратные 5

In [8]:
my_array = np.linspace(5, 100, 20)
print(my_array)

[  5.  10.  15.  20.  25.  30.  35.  40.  45.  50.  55.  60.  65.  70.
  75.  80.  85.  90.  95. 100.]


#### 2. Из полученного массива получить массив с кратными 10 числами

In [9]:
array = my_array[1:20:2]
print(array)

[ 10.  20.  30.  40.  50.  60.  70.  80.  90. 100.]


#### 3. Получить двумерный массив с двумя столбцами

In [10]:
array = array.reshape(5, 2)
print(array)

[[ 10.  20.]
 [ 30.  40.]
 [ 50.  60.]
 [ 70.  80.]
 [ 90. 100.]]


#### 4. Получите набор my_second_array


In [11]:
my_second_array = array - np.full((5, 2), 5)
print(my_second_array)

[[ 5. 15.]
 [25. 35.]
 [45. 55.]
 [65. 75.]
 [85. 95.]]


#### 5. Соедините массивы горизонтально

In [12]:
result = np.concatenate((array, my_second_array), 1)
print(result)

[[ 10.  20.   5.  15.]
 [ 30.  40.  25.  35.]
 [ 50.  60.  45.  55.]
 [ 70.  80.  65.  75.]
 [ 90. 100.  85.  95.]]


#### 6. Разделите 2 и 3 строку на 4 верхних строки my_array. Используйте .T для транспонирования матрицы

In [13]:
result[1:3] /= my_array[:4].T
print(result)

[[ 10.          20.           5.          15.        ]
 [  6.           4.           1.66666667   1.75      ]
 [ 10.           6.           3.           2.75      ]
 [ 70.          80.          65.          75.        ]
 [ 90.         100.          85.          95.        ]]


#### 7. Оставить только те значения, которые не изменились после деления

In [14]:
con = np.full([5, 4], True)
con[1:3].fill(False)
result2 = result[np.where(con)]
print(result2)

[ 10.  20.   5.  15.  70.  80.  65.  75.  90. 100.  85.  95.]


#### 8. Снова получить трехстрочный двумерный массив и поменять местами первую и третью строки

In [15]:
result2 = result2.reshape(3, 4)[::-1]
print(result2)

[[ 90. 100.  85.  95.]
 [ 70.  80.  65.  75.]
 [ 10.  20.   5.  15.]]


#### 9. Получить среднее, медиану и стандартное отклонение для каждой строки

In [16]:
print(np.mean(result2, 1))
print(np.median(result2, 1))
print(np.std(result2, 1))

[92.5 72.5 12.5]
[92.5 72.5 12.5]
[5.59016994 5.59016994 5.59016994]


Среднее
array([92.5, 72.5, 12.5])

Медиана
array([92.5, 72.5, 12.5])

Стандартное отклонение
array([5.59016994, 5.59016994, 5.59016994])

#### 10. Сделать нормализацию на максимальное значение вдоль строк

In [17]:
arrn=(result2.T / result2.max(axis=1)).T
print(arrn)

[[0.9    1.     0.85   0.95  ]
 [0.875  1.     0.8125 0.9375]
 [0.5    1.     0.25   0.75  ]]


### Задача: найти число Пи
![Image of numpy](http://mathfaculty.fullerton.edu/mathews/n2003/montecarlopi/MonteCarloPiMod/Images/MonteCarloPiMod_gr_25.gif)

Внимательно посмотрите на картинку!

Используя методы np.random.rand, np.sqrt, np.sum, np.where, np.ndarray.shape получите вероятностную оценку значения $\pi$ по формуле:
$$\pi = 4*N_{in}/N_{all}$$
Для этого:
* получите $N_{all}$ случайных точек (у каждой точки 2 значения). Все значения от 0 до 1.
* посчитайте радиус от начала координат до точки.
* $N_{in}$ - значения внутри четверти круга с радиусом 1.
* количество точек $N_{all}$ можно варьировать, попробуйте значения от $10$ до $1000000$

In [18]:
Nall = 100
tochki = np.random.rand(Nall, 2)
rad = np.sqrt(np.sum(tochki*tochki, 1))
Nin = np.where(rad <= 1)[0].shape[0]
Pi = 4 * Nin / Nall
print(Pi)

2.8


In [19]:
Nall = 1000000
tochki = np.random.rand(Nall, 2)
rad = np.sqrt(np.sum(tochki*tochki, 1))
Nin = np.where(rad <= 1)[0].shape[0]
Pi = 4 * Nin / Nall
print(Pi)

3.141568


$\pi = 3.1415$

## Pandas
![Image of numpy](https://i.ytimg.com/vi/F6kmIpWWEdU/maxresdefault.jpg)
Докмунтация Pandas: https://pandas.pydata.org/pandas-docs/stable/

### Возможности

In [20]:
import pandas as pd

df = pd.DataFrame({'int_col' : [1,2,6,8,-1], 'float_col' : [0.1, 0.2,0.2,10.1,None], 'str_col' : ['a','b',None,'c','a']})

In [21]:
df

int_col  float_col str_col
0        1        0.1       a
1        2        0.2       b
2        6        0.2    None
3        8       10.1       c
4       -1        NaN       a

In [22]:
# индексация
df.loc[:,['float_col','int_col']]

float_col  int_col
0        0.1        1
1        0.2        2
2        0.2        6
3       10.1        8
4        NaN       -1

In [23]:
# загрузка csv таблицы

df = pd.read_csv('./data/churn.csv')
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [24]:
df.shape

(3333, 21)

In [25]:
df.columns

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')

In [26]:
# описание переменных можно получить методом info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
state                     3333 non-null object
account length            3333 non-null int64
area code                 3333 non-null int64
phone number              3333 non-null object
international plan        3333 non-null object
voice mail plan           3333 non-null object
number vmail messages     3333 non-null int64
total day minutes         3333 non-null float64
total day calls           3333 non-null int64
total day charge          3333 non-null float64
total eve minutes         3333 non-null float64
total eve calls           3333 non-null int64
total eve charge          3333 non-null float64
total night minutes       3333 non-null float64
total night calls         3333 non-null int64
total night charge        3333 non-null float64
total intl minutes        3333 non-null float64
total intl calls          3333 non-null int64
total intl charge         3333 non-null float64

In [27]:
# получить быстро различные статистики можно получить методом describe
df.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [28]:
# статистики по другим переменным
df.describe(include=['object', 'bool'])

state phone number international plan voice mail plan  churn
count   3333         3333               3333            3333   3333
unique    51         3333                  2               2      2
top       WV     385-4715                 no              no  False
freq     106            1               3010            2411   2850

In [29]:
# сортировка
df = df.sort_values(by='total day charge', ascending=False)
df.head()

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   

      total day calls  total day charge  ...  total eve calls  \
365                75             59.64  ...               94   
985                55             58.96  ...               79   
2594               81             58.70  ...              106   
156               120             57.36  ...              116   
605                77             57.04  ...              109   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   

      total intl charge  customer service calls  churn  
365                2.73                       1   True  
985                3.59                       1   True  
2594               3.19                       1   True  
156                4.27                       0   True  
605                3.43                       2   True  

[5 rows x 21 columns]

In [30]:
# индексирование iloc
df.iloc[:15]

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   
975     DE             129        510     332-6181                 no   
15      NY             161        415     351-7269                 no   
1679    OK              82        510     393-4823                 no   
2028    SD              93        510     408-4836                 no   
197     TX             208        510     378-3625                 no   
2412    WI              89        415     373-4264                 no   
2824    OR              81        415     406-4100                 no   
832     SC              87        408     335-1874                 no   
2629    NH             141        415     402-3370                 no   
3132    NC             161        415     394-5489                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   
975               no                      0              334.3   
15                no                      0              332.9   
1679              no                      0              329.8   
2028              no                      0              328.1   
197               no                      0              326.5   
2412              no                      0              326.3   
2824              no                      0              324.7   
832               no                      0              322.5   
2629             yes                     32              322.4   
3132              no                      0              322.3   

      total day calls  total day charge  ...  total eve calls  \
365                75             59.64  ...               94   
985                55             58.96  ...               79   
2594               81             58.70  ...              106   
156               120             57.36  ...              116   
605                77             57.04  ...              109   
975               118             56.83  ...              104   
15                 67             56.59  ...               97   
1679               73             56.07  ...              120   
2028              106             55.78  ...               89   
197                67             55.51  ...              113   
2412              112             55.47  ...              110   
2824               48             55.20  ...               82   
832               106             54.83  ...               93   
2629               92             54.81  ...              107   
3132              100             54.79  ...              135   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   
975              16.33                191.0                 83   
15               27.01                160.6                128   
1679             17.71                267.1                102   
2028             12.89                303.5                114   
197              14.99                181.7                102   
2412        

In [31]:
# индексирование loc
df.loc[:15]

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   
975     DE             129        510     332-6181                 no   
15      NY             161        415     351-7269                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   
975               no                      0              334.3   
15                no                      0              332.9   

      total day calls  total day charge  ...  total eve calls  \
365                75             59.64  ...               94   
985                55             58.96  ...               79   
2594               81             58.70  ...              106   
156               120             57.36  ...              116   
605                77             57.04  ...              109   
975               118             56.83  ...              104   
15                 67             56.59  ...               97   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   
975              16.33                191.0                 83   
15               27.01                160.6                128   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   
975                 8.59                10.4                 6   
15                  7.23                 5.4                 9   

      total intl charge  customer service calls  churn  
365                2.73                       1   True  
985                3.59                       1   True  
2594               3.19                       1   True  
156                4.27                       0   True  
605                3.43                       2   True  
975                2.81                       0   True  
15                 1.46                       4   True  

[7 rows x 21 columns]

In [32]:
# сортировка по разным переменным
df.sort_values(by=['account length', 'total day charge'],
        ascending=[True, False]).head()

state  account length  area code phone number international plan  \
1838    IA               1        408     331-2144                 no   
1490    TN               1        415     335-5591                 no   
1409    NJ               1        415     420-6780                 no   
2960    TX               1        415     396-4254                 no   
2475    AK               1        408     373-1028                 no   

     voice mail plan  number vmail messages  total day minutes  \
1838             yes                     26              208.0   
1490              no                      0              196.1   
1409             yes                     30              183.1   
2960              no                      0              182.1   
2475              no                      0              175.2   

      total day calls  total day charge  ...  total eve calls  \
1838              115             35.36  ...              113   
1490              107             33.34  ...               82   
1409               95             31.13  ...              110   
2960              106             30.96  ...              106   
2475               74             29.78  ...               79   

      total eve charge  total night minutes  total night calls  \
1838             15.73                177.7                144   
1490             25.20                211.5                 91   
1409             19.77                248.3                110   
2960             11.47                152.3                 75   
2475             12.89                230.5                109   

      total night charge  total intl minutes  total intl calls  \
1838                8.00                 8.1                 9   
1490                9.52                 7.0                 2   
1409               11.17                 8.4                 2   
2960                6.85                10.0                 3   
2475               10.37                 5.3                 3   

      total intl charge  customer service calls  churn  
1838               2.19                       1  False  
1490               1.89                       1  False  
1409               2.27                       0  False  
2960               2.70                       5   True  
2475               1.43                       1  False  

[5 rows x 21 columns]

In [33]:
# отбор по условию и вывод средних значений
print(df[df['account length'] == 1].mean())
# отбор по нескольким условиям и получение максимального значения одного признака
print(df[(df['account length'] == 1) & (df['international plan'] == 'no')]['total intl minutes'].max())

account length              1.00000
area code                 412.37500
number vmail messages      10.25000
total day minutes         169.96250
total day calls            98.12500
total day charge           28.89500
total eve minutes         190.27500
total eve calls            87.50000
total eve charge           16.17375
total night minutes       185.75000
total night calls         100.25000
total night charge          8.35750
total intl minutes          8.28750
total intl calls            3.25000
total intl charge           2.23875
customer service calls      1.25000
churn                       0.12500
dtype: float64
13.8


In [34]:
# группирование осуществляется методом groupby

columns_to_show = ['total day minutes', 'total eve minutes', 
                   'total night minutes']

df.groupby(['churn'])[columns_to_show].describe(percentiles=[])

total day minutes                                            \
                  count        mean        std  min    50%    max   
churn                                                               
False            2850.0  175.175754  50.181655  0.0  177.2  315.6   
True              483.0  206.914079  68.997792  0.0  217.6  350.8   

      total eve minutes                                             \
                  count        mean        std   min    50%    max   
churn                                                                
False            2850.0  199.043298  50.292175   0.0  199.6  361.8   
True              483.0  212.410145  51.728910  70.9  211.3  363.7   

      total night minutes                                              
                    count        mean        std   min     50%    max  
churn                                                                  
False              2850.0  200.133193  51.105032  23.2  200.25  395.0  
True                483.0  205.231677  47.132825  47.4  204.80  354.9

### Задачи

In [35]:
data = pd.read_csv('./data/adult.txt', sep=', ')

C:\Users\Acer\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


#### Исследуйте данные

In [36]:
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

#### 1. Сколько в датасете мужчин, сколько женщин? (sex, male, female) ТЗ: =2 строки

In [37]:
print('мужчин:', len(data[data['sex'] == 'Male']), 
      '\nженщин:', len(data[data['sex'] == 'Female']))

мужчин: 21790 
женщин: 10771


#### 2. Какой средний возраст у мужчин, какой у женщин? ТЗ: =2 строки

In [38]:
data.groupby('sex')['age'].mean()

sex
Female    36.858230
Male      39.433547
Name: age, dtype: float64

#### 3. Какой процент из общей выборки составляют представители Кубы (native-country)? ТЗ: =1 строка

In [39]:
print(data[data['native-country'] == 'Cuba'].shape[0] / data.shape[0] * 100, '%')

0.29176008107859097 %


#### 4-5. Посчитайте среднее значение и стандартное отклонение возраста (age) для двух групп (salary): кто зарабатывает меньше 50K, кто зарабатывает больше 50К в год. ТЗ: =2 строки

In [40]:
print('среднее значение:       <=50K:', data[data['salary'] == '<=50K'].mean()['age'], ' ; >50K:', data[data['salary'] == '>50K'].mean()['age'])
print('стандартное отклонение: <=50K:', data[data['salary'] == '<=50K'].std( )['age'], '; >50K:',  data[data['salary'] == '>50K'].std( )['age'])

среднее значение:       <=50K: 36.78373786407767  ; >50K: 44.24984058155847
стандартное отклонение: <=50K: 14.02008849082488 ; >50K: 10.519027719851826


#### 6. Какой процент людей, которые зарабывают больше 50К имеют высшее образование (признак education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate feature)? Воспользуйтесь pd.unique() и .isin(). ТЗ: <=5 строк

In [41]:
feat = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters', 'Doctorate']
print(data[(data['education'].isin(feat)) & (data['salary'] == '>50K')].shape[0] / data.shape[0] * 100, '%')

13.927704923067473 %


#### 7. Покажите статистики по возрасту (age) для представителей разных рас (race) и гендеров (sex). Используйте groupby() и describe() методы. ТЗ: =1 строка

In [42]:
data.groupby(['sex', 'race'])['age'].describe()

count       mean        std   min   25%   50%  \
sex    race                                                                  
Female Amer-Indian-Eskimo    119.0  37.117647  13.114991  17.0  27.0  36.0   
       Asian-Pac-Islander    346.0  35.089595  12.300845  17.0  25.0  33.0   
       Black                1555.0  37.854019  12.637197  17.0  28.0  37.0   
       Other                 109.0  31.678899  11.631599  17.0  23.0  29.0   
       White                8642.0  36.811618  14.329093  17.0  25.0  35.0   
Male   Amer-Indian-Eskimo    192.0  37.208333  12.049563  17.0  28.0  35.0   
       Asian-Pac-Islander    693.0  39.073593  12.883944  18.0  29.0  37.0   
       Black                1569.0  37.682600  12.882612  17.0  27.0  36.0   
       Other                 162.0  34.654321  11.355531  17.0  26.0  32.0   
       White               19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
sex    race                             
Female Amer-Indian-Eskimo  46.00  80.0  
       Asian-Pac-Islander  43.75  75.0  
       Black               46.00  90.0  
       Other               39.00  74.0  
       White               46.00  90.0  
Male   Amer-Indian-Eskimo  45.00  82.0  
       Asian-Pac-Islander  46.00  90.0  
       Black               46.00  90.0  
       Other               42.00  77.0  
       White               49.00  90.0

#### 8. Какие мужчины чаще заработывают больше 50К: женатые или одинокие (marital-status)? Примите, что женатыми являются все, у кого marital-status начинается с Married. ТЗ: <=4 строк

In [43]:
married = data[(data['salary'] == '>50K') & (data['marital-status'].str.contains('^Married'))].shape[0] / data[(data['salary'] == '>50K')].shape[0]
not_married = 1 - married
print('доля женатых   :', married)
print('доля не женатых:', not_married)

доля женатых   : 0.859074097691621
доля не женатых: 0.14092590230837898


#### 9. Какое максимальное значение по количеству рабочих часов в неделю (hours-per-week)? Как много людей, которые столько работают и какой процент среди них зарабатывает больше 50К? ТЗ: <=4 строк

In [46]:
hours = data['hours-per-week'].max()
people = data[data['hours-per-week'] == hours].shape[0]
percent = data[(data['hours-per-week'] == hours) & (data['salary'] == '>50K')].shape[0] / people * 100
print('максимальное значение по количеству рабочих часов в неделю:', hours, '\nколичество людей:', people, '\nпроцент среди них зарабатывает больше 50К:', percent, '%')

максимальное значение по количеству рабочих часов в неделю: 99 
количество людей: 85 
процент среди них зарабатывает больше 50К: 29.411764705882355 %


#### 10. Посчитайте среднее количество рабочих часов в неделю (hours-per-week) для тех кто зарабатывает мало и много (salary) для каждой страны (native-country). ТЗ: =1 строка

In [45]:
data.groupby(['salary', 'native-country'])['hours-per-week'].mean()

salary  native-country            
<=50K   ?                             40.164760
        Cambodia                      41.416667
        Canada                        37.914634
        China                         37.381818
        Columbia                      38.684211
        Cuba                          37.985714
        Dominican-Republic            42.338235
        Ecuador                       38.041667
        El-Salvador                   36.030928
        England                       40.483333
        France                        41.058824
        Germany                       39.139785
        Greece                        41.809524
        Guatemala                     39.360656
        Haiti                         36.325000
        Holand-Netherlands            40.000000
        Honduras                      34.333333
        Hong                          39.142857
        Hungary                       31.300000
        India                         38.233333
     